raft/handler.go

```diff
@@ -881,3 +886,21 @@ func (pm *ProtocolManager) advanceAppliedIndex(index uint64) {
 	pm.appliedIndex = index
 	pm.mu.Unlock()
 }
+
+func (pm *ProtocolManager) updateLeader(leader uint64) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+
+	pm.leader = uint16(leader)
+}
+
+func (pm *ProtocolManager) LeaderAddress() *Address {
+	pm.mu.RLock()
+	defer pm.mu.RUnlock()
+
+	leaderAddress := pm.address
```

Comment:
` leaderAddress := pm.address`
@trung @bts Is it possible because of this line for the local address to be reported as the leader ?

Reply:
as the `pm.peers` doesn't contain the current node so that above statement makes an assumption the leader is the current node being evaluated.

Then the following `if` condition would override if the leader is one of the peers

Let me know if i understand your question wrongly

Reply:
would it be better to keep track of the leader explicitly? leader/follower state change is notified in handleRoleChange function in handler.go

Reply:
This implementation currently implicitly assumes there there is always a leader, which is not the case.

Reply:
that's true. Simple test with 2 nodes showing that leader would never be selected.

Reply:
@nadeemajl role change event doesn't tell you who the current leader is